pydantic/_internal/_namespace_utils.py

```diff
@@ -0,0 +1,237 @@
+from __future__ import annotations
+
+import sys
+from collections.abc import Generator, Mapping
+from contextlib import contextmanager
+from functools import cached_property
+from typing import Any, Iterator, NamedTuple, TypeAlias, TypeVar
+
+GlobalsNamespace: TypeAlias = dict[str, Any]
+"""A global namespace.
+
+In most cases, this is a reference to the `__dict__` attribute of a module.
+This namespace type is expected as the `globals` argument during annotations evaluation.
+"""
+
+MappingNamespace: TypeAlias = Mapping[str, Any]
+"""Any kind of namespace.
+
+In most cases, this is a local namespace (e.g. the `__dict__` attribute of a class,
+the [`f_locals`][frame.f_locals] attribute of a frame object).
+This namespace type is expected as the `locals` argument during annotations evaluation.
+"""
+
+
+class NamespacesTuple(NamedTuple):
+    """A tuple of globals and locals to be used during annotations evaluation.
+
+    This datastructure is defined as a named tuple so that it can easily be unpacked:
+
+    ```python lint="skip" test="skip"
+    def eval_type(typ: type[Any], ns: NamespacesTuple) -> None:
+        return eval(typ, *ns)
+    ```
+    """
+
+    globals: GlobalsNamespace
+    """The namespace to be used as the `globals` argument during annotations evaluation."""
+
+    locals: MappingNamespace
```

Comment:
Why this isnt always `LazyLocalNamespace`? Is something actually allowed to initialize this with any Mapping? There is some empties somewhere in the PR but it could be also then just empty LazyLocalNamespace

Reply:
I wanted to have `LazyLocalNamespace` as an implementation detail, as this gives us more flexibility, especially if in the future we expose the `GenerateSchema` class to Pydantic users (the `types_namespace` property exposes `NamespaceTuple`).